Feature/1457 add shorcuts for pause breaks

evgenyneu opened the pull request:
<!--

Have you read Code of Conduct? By filing an Pull Request, you are expected to comply with it, including treating everyone with respect: https://github.com/hovancik/stretchly/blob/master/CODE_OF_CONDUCT.md

-->

Issue: #1457
<!-- Link to relevant issue. All PRs should be associated with an issue -->

### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.

<!-- If there’s a checkbox you can’t complete for any reason, that's okay, just explain in detail why you weren’t able to do so. -->

- [x]  issue was opened to discuss proposed changes before starting implementation. It is important do discuss changes before implementing them (Why should we add it? How should it work? How should it look? Where will it be? ...).
- [x]  during development, `node` version specified in `package.json` was used (ie using [nvm](https://github.com/creationix/nvm)).
- [x]  package versions and package-lock.json were not changed (`npm install --no-save`).
- [x]  app version number was not changed.
- [x]  all new code has tests to ensure against regressions.
- [x] `npm run lint` reports no offenses.
- [x] `npm run test` is error-free.
- [x]  README and CHANGELOG were updated accordingly.
- []  after PR is approved, all commits in it are [squashed](https://gitbetter.substack.com/p/how-to-squash-git-commits)

### Description of the Change

* Added ability to set keyboard shortcuts to pause the breaks for the period of time (30 min, 1 hr etc.)
* Moved the code for keyboard chortcut registration from `main.js` into `utils/breakShortcuts.js`.

### Verification Process

* Open About tab in the Preferences screen.
* Click Ctrl-D to see the path to preferences JSON
* Set the keyboard shortcuts:

```
"pauseBreaksFor30MinutesShortcut": "CmdOrCtrl+Alt+0",
"pauseBreaksFor1HourShortcut": "CmdOrCtrl+Alt+1",
"pauseBreaksFor2HoursShortcut": "CmdOrCtrl+Alt+2",
"pauseBreaksFor5HoursShortcut": "CmdOrCtrl+Alt+5",
"pauseBreaksUntilMorningShortcut": "CmdOrCtrl+Alt+M",
```

* Restart the app.
* Use the keyboard shortcuts and see that the breaks are paused: the app icon changes to Pause and when you click the text shows the correct pause duration.

* Since I moved the code for other break shortcuts into, I also tested these shortcuts:
```
"pauseBreaksToggleShortcut": "CmdOrCtrl+Alt+T",
"skipToNextScheduledBreakShortcut": "CmdOrCtrl+Alt+B",
"skipToNextMiniBreakShortcut": "CmdOrCtrl+Alt+S",
"skipToNextLongBreakShortcut": "CmdOrCtrl+Alt+L",
"resetBreaksShortcut": "CmdOrCtrl+Alt+R",
```

* I tested it on Mac, Windows and Ubuntu.

evgenyneu committed:
Fix test

Update timezone in unit tests to make test pass.
Test failed when I ran them in
Melbourne because the hardcoded time stamps
were created in my local timezone (UTC+10)
while the test expectation
assume the times are in Amsterdam's time zone (UTC+2)

evgenyneu committed:
Add keyboard shortcuts for pause breaks

evgenyneu committed:
Document

evgenyneu committed:
Changelog

hovancik commented on app/main.js:
```diff
@@ -1315,26 +1317,31 @@ function getTrayMenuTemplate () {
       submenu: [
         {
           label: i18next.t('utils.minutes', { count: 30 }),
+          accelerator: settings.get('pauseBreaksFor30MinutesShortcut'),
```
I guess not that important but returns `[18960:0723/192752.999383:WARNING:accelerator_util.cc(65)]  doesn't contain a valid key` when no value set.

evgenyneu replied:
Well spotted. How do I reproduce this error? I tried to start the app (`npm run start`) with empty string `""` values for the shortcuts but I don't see this error in the terminal.

hovancik replied:
Try with `npm run dev`, it gives more debug info.

evgenyneu replied:
Fixed, the warning only appeared in Ubuntu for me but not in Windows or Mac.

hovancik commented on app/utils/pauseBreaksShortcut.js:
```diff
@@ -0,0 +1,44 @@
+const { UntilMorning } = require('./untilMorning')
```
Could you please what do you think about registering `pauseBreaksToggleShortcut`? Haven't had time think myself, but it hits me we have one shortcut left in `main.js` regarding to pausing.

Or even put all shortcuts here? (like skipToNextScheduledBreakShortcut, skipToNextMiniBreakShortcut and such).

evgenyneu replied:
Done

hovancik commented:
I've made some comments, let me know what you think :) 

I can test then on Windows as well.

evgenyneu committed:
Use object parameter

evgenyneu committed:
Move pauseBreaksToggleShortcut out of main.js

evgenyneu committed:
Move skipToNextScheduledBreakShortcut out of main.js

evgenyneu committed:
More tests for registerPauseBreaksShortcuts

evgenyneu committed:
Move skipToNextMiniBreakShortcut out of main.js

evgenyneu committed:
Move skipToNextLongBreakShortcut out of main.js

evgenyneu committed:
Move resetBreaksShortcut out of main.js

evgenyneu committed:
Refactor

evgenyneu committed:
Refactor

evgenyneu committed:
Rename break shortcuts files for clarity

evgenyneu committed:
Fix empty menu accelerator warning

The warning "doesn't contain a valid key"
was shown in console logs when pause break shortcut was empty.
The warning only appread in  Linux but not in Windows or Mac.

evgenyneu commented:
I've made the improvements you suggested. Also, I managed to install the dev environment on Windows, and tested the shortcuts on all platforms. Let me know if you want to change/improve anything. :D

hovancik reviewed (approved):
Many thanks, looks good :)

hovancik commented:
I can squash and merge, unless you want to add yourself to the readme :)

evgenyneu commented:
Thanks, squash is ok, I don't need readme mention :D
